File: python/is_interleave.py
class Solution(object):
    # O(n * m) time | O(n * m) space
    def isInterleave(self, s1, s2, s3):
        """
        :type s1: str
        :type s2: str
        :type s3: str
        :rtype: bool
        """
        if len(s1) + len(s2) != len(s3):
            return False
        
        dp = [ [ None ] * (len(s1) + 1) for _ in range(len(s2) + 1)  ]

        def dfs(i, j):
            if i + j == len(s3):
                
                return True
            if dp[i][j]:
                return dp[i][j]
            
            k = i + j
            found = False
            if i < len(s1) and s1[i] == s3[k]:
                found = dfs(i + 1, j)
                if found:
                    return True
                
            if j < len(s2) and s2[j] == s3[k]:
                found = dfs(i, j + 1)
            
            dp[i][j] = found
            return found

        return dfs(0, 0)
    

solution = Solution()
s1 = "aabcc"
s2 = "dbbca"
s3 = "aadbbcbcac"
print(solution.isInterleave(s1, s2, s3))

File: src/main/java/june_2025/IsInterleave.java
package june_2025;

public class IsInterleave {

    public static void main(String[] args) {
        // s1 = "aabcc", s2 = "dbbca", s3 = "aadbbcbcac"
        String s1 = "aabcc";
        String s2 = "dbbca";
        String s3 = "aadbbcbcac";

        IsInterleave isInterleave = new IsInterleave();
        boolean result = isInterleave.isInterleave(s1, s2, s3);
        System.out.println(result);
    }

    // O(n*m) time | o(n*m) space
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) {
            return false;
        }
        Boolean[][] dp = new Boolean[s1.length() + 1][s2.length() + 1];
        return dfs(s1, s2, s3, 0, 0, dp);
    }

    private boolean dfs(String s1, String s2, String s3, int i, int j, Boolean[][] dp) {
        if (i + j == s3.length()) {
            return true;
        }
        if (dp[i][j] != null) {
            return dp[i][j];
        }
        int k = i + j;

        boolean res = false;
        if (i < s1.length() && s1.charAt(i) == s3.charAt(k)) {
            res = dfs(s1, s2, s3, i + 1, j, dp);
        }
        if (!res && j < s2.length() && s2.charAt(j) == s3.charAt(k)) {
            res = dfs(s1, s2, s3, i, j + 1, dp);
        }
        dp[i][j] = res;
        return res;
    }

}
